File: 0105-construct-binary-tree-from-preorder-and-inorder-traversal/0105-construct-binary-tree-from-preorder-and-inorder-traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr, right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* buildTreeHelper(vector<int>& preorder, int preStart, int preEnd, vector<int>& inorder, int inStart, int inEnd, unordered_map<int,int>& inorderNodeLoc) {
        if (preStart > preEnd || inStart > inEnd) return nullptr;

        TreeNode *root = new TreeNode(preorder[preStart]);
        int rootLocInInorder = inorderNodeLoc[root->val];
        int nodesOnLeft = rootLocInInorder - inStart;

        root->left = buildTreeHelper(preorder, preStart + 1, preStart + nodesOnLeft, inorder, inStart, rootLocInInorder - 1, inorderNodeLoc);
        root->right = buildTreeHelper(preorder, preStart + nodesOnLeft + 1, preEnd, inorder, rootLocInInorder + 1, inEnd, inorderNodeLoc);

        return root;
    }

    TreeNode* buildTree(vector<int>& preorder, vector<int>& inorder) {
        unordered_map<int, int> inorderNodeLoc;
        for (int i = 0; i < inorder.size(); i++) {
            inorderNodeLoc[inorder[i]] = i;
        }

        TreeNode* root = buildTreeHelper(preorder, 0, preorder.size() - 1, inorder, 0, inorder.size() - 1, inorderNodeLoc);
        return root;
    }
};

File: 0105-construct-binary-tree-from-preorder-and-inorder-traversal/0105-construct-binary-tree-from-preorder-and-inorder-traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def buildTree(self, preorder: List[int], inorder: List[int]) -> Optional[TreeNode]:
        inorderLoc = {}

        for i in range(len(inorder)):
            inorderLoc[inorder[i]] = i
        
        preorder = collections.deque(preorder)

        def build(start,end):
            if start > end:
                return None

            root = TreeNode(preorder.popleft())

            mid = inorderLoc[root.val]

            root.left = build(start, mid-1)
            root.right = build(mid+1 , end)

            return root

        return build(0, len(preorder)-1)
        

